// echo-plugin/index.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// @ts-check
const createPlugin = require('../../../../_plugin');
const path = require('path');

/**
 * @augments {TemplateStringLanguageService}
 */
class TestStringLanguageService {
    getDefinitionAndBoundSpan(context) {
        return {
            definitions: context.text === 'no_definition' ? [] : [{
                name: 'foo',
                kind: 'string',
                containerKind: '',
                containerName: '',
                fileName: 'main.ts',
                textSpan: {
                    start: 0,
                    length: context.node.getFullWidth(),
                }
            }],
            textSpan: {
                start: 0,
                length: context.node.getFullWidth(),
            }
        };
    }
}

module.exports = createPlugin(
    (_log) => {
        return new TestStringLanguageService()
    }, {
        tags: ['test'],
        enableForStringWithSubstitutions: true,
        getSubstitution(text, start, end) {
            return 'x'.repeat(end - start);
        }
    })
