// echo-plugin/index.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// @ts-check
const createPlugin = require('../../../../_plugin');

/**
 * @implements {TemplateLanguageService}
 */
class TestStringLanguageService {
    constructor(log) {
        this.log = log;
    }

    getSignatureHelpItemsAtPosition({ text }, position) {
        const items = [];

        const line = text.split(/\n/g)[position.line];
        const afterParen = line.match(/\((.+?)$/);
        if (afterParen) {
            items.push({
                isVariadic: false,
                prefixDisplayParts: [
                    { text: afterParen[1], kind: 'text' }
                ],
                suffixDisplayParts: [],
                separatorDisplayParts: [],
                parameters: [],
                documentation: [],
                tags: [],
            });
        }
        if (!items.length) {
            return undefined;
        }
        
        return {
            items,
            applicableSpan: { start: 0, length: 1 },
            selectedItemIndex: 0,
            argumentIndex: 0,
            argumentCount: 0,
        }
    }
}

module.exports = createPlugin(
    (_ts, log) => {
        return new TestStringLanguageService(log)
    }, {
        tags: ['test'],
        enableForStringWithSubstitutions: true,
        getSubstitution(text, start, end) {
            return 'x'.repeat(end - start);
        }
    })
