// echo-plugin/index.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// @ts-check
const createPlugin = require('../../../../_plugin');

/**
 * @augments {TemplateStringLanguageService}
 */
class TestStringLanguageService {
    getOutliningSpans({ text }) {
        const spans = [];
        const spanRe = /{[^}]*}/gm;
        let match;
        while ((match = spanRe.exec(text))) {
            spans.push({
                textSpan: {
                    start: match.index,
                    length: match[0].length
                },
                hintSpan: {
                    start: match.index + 1,
                    length: match[0].length - 2
                }
            });
        }
        return spans;
    }
}

module.exports = createPlugin(
    (_log) => {
        return new TestStringLanguageService()
    }, {
        tags: ['test'],
        enableForStringWithSubstitutions: true,
        getSubstitution(text, start, end) {
            return 'x'.repeat(end - start);
        }
    })
